internal/mongo/watch_producer.go

```diff
@@ -31,49 +32,95 @@ func (w *WatchProducer) GetProducer(o ...WatchOption) ChangeEventProducer {
 			pipeline = append(customElements, pipeline...)
 		}
 
-		opts := &options.ChangeStreamOptions{
-			BatchSize:            &config.batchSize,
-			MaxAwaitTime:         &config.maxAwaitTime,
-			StartAtOperationTime: config.startAtOperationTime,
-		}
-		if config.fullDocumentEnabled {
-			opts.SetFullDocument(options.UpdateLookup)
-		}
-		if len(config.resumeAfter) != 0 {
-			opts.SetResumeAfter(config.resumeAfter)
-		}
-
-		cursor, err := w.collection.Watch(ctx, pipeline, opts)
+		cursor, err := w.watch(ctx, pipeline, config, nil)
 		if err != nil {
-			w.logger.Error("Mongo client: An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+			w.logger.Error("Mongo client: An error has occured while trying to watch collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
 			return nil, err
 		}
 
 		var events = make(chan *ChangeEvent)
 
 		go func() {
-			defer cursor.Close(ctx)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
 			defer close(events)
-
 			for {
-				w.sendEvents(ctx, cursor, events)
+				resumeToken, err := w.sendEvents(ctx, cursor, events)
+				w.logger.Error("Mongo client : An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+
+				cursor.Close(ctx)
```

Comment:
why should we open & close a cursor, i don't think it is performant ?

Reply:
Because we need to open a new one each time we retry to watch the collection ;)

Reply:
It should keep the same cursor when no error occurs and open a new cursor only when there is an error and so an attempt to retry ?

Reply:
It is the case, I think you misunderstood something there.